webapp/api/github.py

```diff
@@ -0,0 +1,127 @@
+    """
+
+    def __init__(self):
+        self.client_id = GITHUB_CLIENT_ID
+        self.client_secret = GITHUB_CLIENT_SECRET
+        self.base_url = "https://api.github.com/"
+        self.base_auth_url = "https://github.com/login/oauth/"
+        self.session = api.requests.Session()
+        self.session.headers["Accept"] = "application/json"
+
+    def authorize(self, scope=None, state=None):
+        """
+        Redirect to Github and request access to a user's data.
+        """
+        params = {
+            "client_id": self.client_id,
+        }
+
+        if scope:
+            params.update({"scope": scope})
+
+        if state:
+            params.update({"state": state})
+
+        auth_url = self.base_auth_url + "authorize?" + urlencode(params)
+
+        return flask.redirect(auth_url)
+
+    def verify_auth(self, code):
+        """
+        Handles response after the redirect to Github. This response
+        determines if the user has allowed this application access. If we
+        were then we send a POST request for the access_key used to
+        authenticate requests to Github.
+        """
+        params = {
+            "code": code,
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+        }
+
+        response = self.request(
+            "POST", self.base_auth_url, "access_token", params
+        )
+
+        try:
+            data = response.json()
+        except ValueError:
+            pass
+
+        if not response.ok or "error" in data or "access_token" not in data:
+            raise GithubAuthException("Authentication token exchange failed.")
+
+        return data["access_token"]
+
+    def request(
+        self, method, base_url, resource, params={}, access_token=None
+    ):
+        """
+        Makes a raw HTTP request and returns the response.
+        """
+        url = base_url + resource
+
+        if access_token is None:
+            access_token = flask.session.get("github_auth_secret")
+
+        if access_token:
+            params.update({"access_token": access_token})
+            self.session.headers["Authorization"] = f"token {access_token}"
+
+        return self.session.request(method, url, json=params)
+
+    def get_resource(self, resource, params=None, access_token=None):
+        """
+        Makes a raw HTTP GET request and returns the response.
+        """
+        data = None
+
+        response = self.request(
+            "GET", self.base_url, resource, params, access_token
+        )
+
+        if response.ok:
+            data = response.json()
+
+        return response, data
+
+    def get_user(self):
+        """
+        Sets the user to the currently authorized user if it is not already and
+        returns it. The user object is the dictionary response from the API.
+        """
+        try:
+            resp, user = self.get_resource("user")
+        except Exception:
```

Comment:
I'd rather not have such a broad try in here. If we don't care about potential failures we can do it on a case by case basis for the specific function call?